pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -1120,6 +1127,74 @@ func (r ReconcileHyperConverged) amendBadDefaults(req *common.HcoRequest) (bool,
 	return modified, nil
 }
 
+func (r ReconcileHyperConverged) applyUpgradePatches(req *common.HcoRequest) (bool, error) {
+	modified := false
+
+	knownHcoVersion, _ := req.Instance.Status.GetVersion(hcoVersionName)
+	if knownHcoVersion == "" {
+		knownHcoVersion = "0.0.0"
+	}
+
+	err := readUpgradePatchesFromFile(req)
+	if err != nil {
+		return false, err
+	}
+	err = validateUpgradePatches(req)
+	if err != nil {
+		return false, err
+	}
+
+	knownHcoSV, err := semver.Parse(knownHcoVersion)
+	if err != nil {
+		req.Logger.Error(err, "Error!")
+		return false, err
+	}
+
+	for _, p := range hcoUpgradeChanges.HCOCRPatchList {
+		m, err := r.applyUpgradePatch(req, knownHcoSV, p)
+		if err != nil {
+			return false, err
+		}
+		modified = modified || m
+	}
+	return modified, nil
+}
+
+func (r ReconcileHyperConverged) applyUpgradePatch(req *common.HcoRequest, knownHcoSV semver.Version, p hcoCRPatch) (bool, error) {
+	modified := false
+	affectedRange, err := semver.ParseRange(p.SemverRange)
+	if err != nil {
+		return modified, err
+	}
+	if affectedRange(knownHcoSV) {
+		patches, err := jsonpatch.DecodePatch([]byte(p.JSONPatch))
+		if err != nil {
+			return modified, err
+		}
+		req.Logger.Info("applying upgrade patch", "knownHcoSV", knownHcoSV, "affectedRange", affectedRange, "patches", patches)
+		specBytes, err := json.Marshal(req.Instance)
+		if err != nil {
+			return modified, err
+		}
```

Comment:
Please take it outside the loop

Reply:
I think we should keep it in to ensure we are applying each patch on top of the previous ones

Reply:
Marshalling and unmarshaling costs cpu and memory. I think this is notvsomething to do in a loop. Also, here you also do it one on top the other, but on a json representation rather than cr.

Reply:
done